workflow/sync/sync_manager.go

```diff
@@ -113,102 +111,146 @@ func (cm *Manager) Initialize(wfs []wfv1.Workflow) {
 	if err != nil {
-		return false, false, "", fmt.Errorf("requested configuration is invalid: %w", err)
+		return false, false, "", failedLockName, fmt.Errorf("requested configuration is invalid: %w", err)
 	}
+	holderKey := getHolderKey(wf, nodeName)
 
-	lockKey := syncLockName.EncodeName()
-	lock, found := cm.syncLockMap[lockKey]
-	if !found {
-		switch syncLockRef.GetType() {
-		case wfv1.SynchronizationTypeSemaphore:
-			lock, err = cm.initializeSemaphore(lockKey)
-		case wfv1.SynchronizationTypeMutex:
-			lock = cm.initializeMutex(lockKey)
-		default:
-			return false, false, "", fmt.Errorf("unknown Synchronization Type")
-		}
+	var msg string
+	lockKeys := make([]string, len(syncItems))
+	allAcquirable := true
+	for i, syncItem := range syncItems {
+		syncLockName, err := getLockName(syncItem, wf.Namespace)
 		if err != nil {
-			return false, false, "", err
+			return false, false, "", failedLockName, fmt.Errorf("requested configuration is invalid: %w", err)
 		}
-		cm.syncLockMap[lockKey] = lock
+		lockKeys[i] = syncLockName.encodeName()
 	}
+	for i, lockKey := range lockKeys {
+		lock, found := sm.syncLockMap[lockKey]
+		if !found {
+			var err error
+			switch syncItems[i].getType() {
+			case wfv1.SynchronizationTypeSemaphore:
+				lock, err = sm.initializeSemaphore(lockKey)
+			case wfv1.SynchronizationTypeMutex:
+				lock = sm.initializeMutex(lockKey)
+			default:
+				return false, false, "", failedLockName, fmt.Errorf("unknown Synchronization Type")
+			}
+			if err != nil {
+				return false, false, "", failedLockName, err
+			}
+			sm.syncLockMap[lockKey] = lock
+		}
 
-	if syncLockRef.GetType() == wfv1.SynchronizationTypeSemaphore {
-		err := cm.checkAndUpdateSemaphoreSize(lock)
-		if err != nil {
-			return false, false, "", err
+		if syncItems[i].getType() == wfv1.SynchronizationTypeSemaphore {
+			err := sm.checkAndUpdateSemaphoreSize(lock)
+			if err != nil {
+				return false, false, "", "", err
+			}
 		}
-	}
 
-	holderKey := getHolderKey(wf, nodeName)
-	var priority int32
-	if wf.Spec.Priority != nil {
-		priority = *wf.Spec.Priority
-	} else {
-		priority = 0
-	}
-	creationTime := wf.CreationTimestamp
-	lock.addToQueue(holderKey, priority, creationTime.Time)
-
-	ensureInit(wf, syncLockRef.GetType())
-	currentHolders := cm.getCurrentLockHolders(lockKey)
-	acquired, msg := lock.tryAcquire(holderKey)
-	if acquired {
-		updated := wf.Status.Synchronization.GetStatus(syncLockRef.GetType()).LockAcquired(holderKey, lockKey, currentHolders)
-		return true, updated, "", nil
+		var priority int32
+		if wf.Spec.Priority != nil {
+			priority = *wf.Spec.Priority
+		} else {
+			priority = 0
+		}
+		creationTime := wf.CreationTimestamp
+		lock.addToQueue(holderKey, priority, creationTime.Time)
+
+		ensureInit(wf, syncItems[i].getType())
+		acquired, already, newMsg := lock.checkAcquire(holderKey)
+		if !acquired && !already {
+			allAcquirable = false
+			if msg == "" {
+				msg = newMsg
+			}
```

Comment:
Is it important that we keep initialising locks when we know we cannot proceed?
Why not `break` instead?

Reply:
Each lock has a separate queue. We want to call addToQueue for all the locks we'd like to acquire.

Pop a break in here and the multiple semaphore tests [TestPriority/Priority wf1 will be able to grab the mutex here](https://github.com/argoproj/argo-workflows/blob/22b64f98998754216e56bbf768fc66f92ecb0f5e/workflow/sync/multiple_test.go#L378) before `wfhigh` violating priority, and [TestMultipleMutexLock/MultipleMutexOrdering wf2 will be able to grab mutex two despite wfall asking for it first](https://github.com/argoproj/argo-workflows/blob/22b64f98998754216e56bbf768fc66f92ecb0f5e/workflow/sync/multiple_test.go#L158) violating ordering.

Reply:
Right okay fair enough in that case.